Description: Fix e1000e hung at stats update
Author: Rouven Spreckels <n3vu0r@qu1x.org>
Origin: IPFire, <http://git.ipfire.org/?p=people/arne_f/ipfire-2.x.git;a=commit;h=fcffac1340f61923786af8a860c12056a9ef3706>
Bug: https://lists.debian.org/debian-kernel/2017/06/msg00128.html
Last-Update: 2019-03-06
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
--- a/drivers/net/ethernet/intel/e1000e/netdev.c
+++ b/drivers/net/ethernet/intel/e1000e/netdev.c
@@ -5940,9 +5940,13 @@ void e1000e_get_stats64(struct net_devic
 			struct rtnl_link_stats64 *stats)
 {
 	struct e1000_adapter *adapter = netdev_priv(netdev);
+	int has_lock = 0;
+
+	if (spin_trylock(&adapter->stats64_lock)) {
+		e1000e_update_stats(adapter);
+		has_lock = 1;
+	}
 
-	spin_lock(&adapter->stats64_lock);
-	e1000e_update_stats(adapter);
 	/* Fill out the OS statistics structure */
 	stats->rx_bytes = adapter->stats.gorc;
 	stats->rx_packets = adapter->stats.gprc;
@@ -5972,7 +5976,7 @@ void e1000e_get_stats64(struct net_devic
 
 	/* Tx Dropped needs to be maintained elsewhere */
 
-	spin_unlock(&adapter->stats64_lock);
+	if (has_lock) spin_unlock(&adapter->stats64_lock);
 }
 
 /**
--- a/drivers/net/ethernet/intel/igb/igb_main.c
+++ b/drivers/net/ethernet/intel/igb/igb_main.c
@@ -6236,10 +6236,11 @@ static void igb_get_stats64(struct net_d
 {
 	struct igb_adapter *adapter = netdev_priv(netdev);
 
-	spin_lock(&adapter->stats64_lock);
-	igb_update_stats(adapter);
+	if (spin_trylock(&adapter->stats64_lock)) {
+		igb_update_stats(adapter);
+		spin_unlock(&adapter->stats64_lock);
+	}
 	memcpy(stats, &adapter->stats64, sizeof(*stats));
-	spin_unlock(&adapter->stats64_lock);
 }
 
 /**
